docs/rfc/rfc-8.adoc

```diff
@@ -0,0 +1,410 @@
+for more details.
+
+=== Supporting cross-ecosystem bridges
+
+The Wormhole bridge will be the first cross-ecosystem bridge to have a minting
+authority for tBTC token on Arbitrum, Polygon, and Optimism.
+
+Other cross-ecosystem bridges may be added in the future given the canonical
+tBTC token capability of delegating the minting authority.
+
+==== Wormhole
+
+The Wormhole bridging process from L1 to L2 (or sidechain) will look as follows:
+
+* There is a tBTC holder on L1. The holder goes to the Wormhole Portal and
+  selects the chain they want to bridge to.
+* The holder submits one transaction to L1. After the transaction is mined, they
+  wait about 15 minutes for the Ethereum block finality.
+* The holder submits one transaction to L2. After that transaction is mined,
+  they have their tBTC on L2.
+
+The Wormhole bridging process from L2 (or sidechain) to L1 will look as follows:
+
+* There is a tBTC holder on L2. That holder goes to the Wormhole portal and
+  selects one of the L2 chains they want to bridge from.
+* The holder submits one transaction to L2. After the transaction is mined, they
+  wait about 15 minutes for the L2 block finality.
+* The holder submits one transaction to L1. After that transaction is mined,
+  they have their tBTC on L1.
+
+What is not immediately obvious is that the token received on L2 may not be the
+canonical token on that L2.
+
+For example, when I crossed the bridge with ETH, I received WETH token at
+address link:https://arbiscan.io/token/0xd8369c2eda18dd6518eabb1f85bd60606deb39ec[`0xd8369c2eda18dd6518eabb1f85bd60606deb39ec`].
+This is not the canonical WETH on Arbitrum. The canonical WETH on Arbitrum is
+link:https://arbiscan.io/token/0x82af49447d8a07e3bd95bd0d56f35241523fbab1[`0x82af49447d8a07e3bd95bd0d56f35241523fbab1`].
+When I crossed the bridge with USDC I received USDC token at
+link:https://arbiscan.io/token/0xc96f2715e2a242d50d1b0bc923dbe1740b8ecf18[`0xc96f2715e2a242d50d1b0bc923dbe1740b8ecf18`]
+which is not the canonical USDC on Arbitrum. The canonical USDC on Arbitrum is
+link:https://arbiscan.io/token/0xff970a61a04b1ca14834a43f5de4533ebddb5cc8[`0xff970a61a04b1ca14834a43f5de4533ebddb5cc8`].
+
+Wormhole cannot magically mint all tokens on Arbitrum or other L2s. Wormhole
+mints its own token representing the bridged asset that may be swapped to the
+canonical representation on the given L2 assuming there is a market for such
+a swap. The overview of liquid markets on each target chain is available in the
+link:https://docs.wormhole.com/wormhole/overview-liquid-markets[Wormhole documentation].
+Not every chain and token pair has a market. 
+
+To fully use the capabilities of the Wormhole portal and make the user
+experience seamless, we will implement a gateway contract having the authority to
+wrap Wormhole-specific tokens into tBTC tokens in a 1:1 ratio. This contract
```

Comment:
Users will be able to send L2 tBTC to ETH using wormhole.  This will make a wormhole wrapped tBTC representation on ETH.  How will users unwrap this to BTC?

Reply:
Can you elaborate more here? Here's the situation I'm imagining:

1) User deposits BTC and mints TBTC on ethereum (via tbtc btc<->eth bridge)
2) User locks up the TBTC on ethereum and mints wormholeL2TBTC (via wormhole eth<->L2 bridge)
3) wormholeL2TBTC gets locked on L2 to mint L2TBTC (via L2 vending machine)

Then ideally we have:

4) User burns L2TBTC to unlock wormholeL2TBTC (via L2 vending machine)
5) User burns wormholeL2TBTC to unlock TBTC (via wormhole eth<->L2 bridge)

But if I'm understanding this correctly, they could *also*

4) User locks L2TBTC to mint wormholeL1TBTC on ethereum (via wormhole eth<->L2 bridge)

So now you have TBTC and wormholeTBTC both on ethereum, which ends up fracturing liquidity

Reply:
> This will make a wormhole wrapped tBTC representation on ETH. How will users unwrap this to BTC?

There should never be Wormhole-wrapped tBTC on Ethereum, right? If it lands on Ethereum it will have been minting by someone else locking tBTC on Ethereum, so it should be unwrappable.

Reply:
L2tBTC when bridged back to ethereum will produce a bridge wrapped version of L2tBTC ("wormholeL2tBTC").  In order to turn L2tBTC into proper ethereum tBTC, the user must use the L2 vending machine first before using the bridge to get L2tBTC.

I imagine users will often forget to use the vending machine when bridging back to Ethereum.  I can think of two options:

a) It would be convenient to allow swapping wormholeL2tBTC to tBTC.  However, there's still tBTC locked on Wormhole that would need to be burned or accounted for somehow.
b) It could also be up to the bridge to special case L2tBTC to always use the vending machine so hopefully the user doesn't end up with wormholeL2tBTC to begin with.  Worst case the user has to redo what they did.

Reply:
> L2tBTC when bridged back to ethereum will produce a bridge wrapped version of L2tBTC ("wormholeL2tBTC"). In order to turn L2tBTC into proper ethereum tBTC, the user must use the L2 vending machine first before using the bridge (L2tBTC -> whtBTC on L2 -> tBTC).

Correct. The proposal assumes - although I see now it is unclear - that the Wormhole contract on L2 can be customized to allow for wrapping the token from wormholeL2tBTC into a canonical L2tBTC when redeeming on L2 and to unwrap the token from the canonical L2tBTC to wormholeL2tBTC when depositing into the L2 contract to bridge back.

I have also mentioned it in [this](https://github.com/keep-network/tbtc-v2/pull/542#discussion_r1114303689) comment.

> I imagine users will often forget to use the vending machine when bridging back to Ethereum. 

Yes, this must happen automatically - either by adding wrap/unwrap functionality to the contract - ideal for the UX. Alternatively, integrated into the Portal but it would mean more than one transaction.

>  It could also be up to the bridge to special case L2tBTC to always use the vending machine so hopefully the user doesn't end up with wormholeL2tBTC to begin with.

Yes, that's the preferred way for this proposal. From the user's perspective, the easiest way is when they never receive wormholeL2tBTC into their wallet. When they redeem on L2, wormholeL2tBTC is wrapped automatically and they receive L2tBTC. When they deposit to bridge back to L1, they deposit L2tBTC and unwrapping happens internally in the Wormhole's bridge contract.

Reply:
@conorpp was kind to provide examples and describe the possible approaches. I summarised our options in b629bcb13b475fe26757b7ee748aa4bdcd2956e3.

Reply:
> Yes, that's the preferred way for this proposal. From the user's perspective, the easiest way is when they never receive wormholeL2tBTC in their wallet. When they redeem on L2, wormholeL2tBTC is wrapped automatically, and they receive L2tBTC. When they deposit to bridge back to L1, they deposit L2tBTC, and unwrapping happens internally in the Wormhole's bridge contract.

This could be the best UX solution for how the gateway works on the user-facing side. The user should receive the canonical token and be unaware of the wrapping and unwrapping from wormholetBTC to L2tBTC, and this should magically happen under the hood. 

I think coercing the user to find a market to swap their non-canonical token into a canonical token is a bit of a UX nightmare, and I also believe that it may decrease the token use and its adoption if we make it too hard for the users to use their bridged tokens on the L2s.